File: EsqExt.cs
using System;
using System.Linq;
using System.Collections.Generic;
using Terrasoft.Common;
using Terrasoft.Core;
using Terrasoft.Core.Entities;


namespace VSK.Base
{
	public class EsqExt<T> : EntitySchemaQuery where T: Entity
	{
		private readonly UserConnection uc;

		public EsqExt(UserConnection userConnection)
			: base(userConnection.EntitySchemaManager, typeof(T).Name)
		{
			uc = userConnection;
		}

		public EsqExt<T> Reset()
		{
			Columns.Clear();
			AddColumn("Id");
			Filters.Clear();
			ResetSelectQuery();
			PrimaryQueryColumn.IsAlwaysSelect = true;
			IgnoreDisplayValues = false;
			UseAdminRights = true;
			return this;
		}

		public EsqExt<T> AddColumn(string columnPath, string alias = null)
		{
			var column = base.AddColumn(columnPath);
			if (alias.IsNotNullOrEmpty()) column.Name = alias;
			return this;
		}

		public EsqExt<T> AddAllColumns()
		{
			AddAllSchemaColumns();
			return this;
		}

		public EsqExt<T> AddColumnOrderedBy(string columnPath, OrderDirection orderDirection, int orderPosition = 1, string alias = null)
		{
			var column = base.AddColumn(columnPath);
			column.OrderDirection = orderDirection;
			column.OrderPosition = orderPosition;
			if (alias.IsNotNullOrEmpty()) column.Name = alias;
			return this;
		}

		public T GetEntity(Guid id)
		{
			RowCount = 1;
			var entity = GetEntity(uc, id);
			return entity != null ? (T)entity : null;
		}

		public T GetFirstOrDefault()
		{
			RowCount = 1;
			var entity = GetEntityCollection(uc).FirstOrDefault();
			return entity != null ? (T)entity : null;
		}

		public IEnumerable<T> GetEntityCollection()
		{
			var entities = GetEntityCollection(uc);
			return entities.Select(e => (T)e);
		}

		public int GetEntityCollectionCount()
		{
			Columns.Clear();
			PrimaryQueryColumn.IsAlwaysSelect = false;
			var countCol = AddColumn(CreateAggregationFunction(AggregationTypeStrict.Count, "Id"));
			var entities = GetEntityCollection(uc);
			return entities.Count > 0 ? entities[0].GetTypedColumnValue<int>(countCol.Name) : 0;
		}

		public IEnumerable<IEnumerable<T>> GetEntityCollectionPageIterator(int entitiesOnPage)
		{
			if (entitiesOnPage <= 0)
				throw new ArgumentOutOfRangeException(nameof(entitiesOnPage));
			throw new NotImplementedException();
		}

		#region Filters
		public EsqExt<T> AddInFilter(string fieldName, params Guid[] values)
		{
			var valuesAsObj = values.Select(v => (object)v);
			Filters.Add(CreateFilterWithParameters(FilterComparisonType.Equal, fieldName, valuesAsObj));
			return this;
		}
		public EsqExt<T> AddNotInFilter(string fieldName, params Guid[] values)
		{
			var valuesAsObj = values.Select(v => (object)v);
			Filters.Add(CreateFilterWithParameters(FilterComparisonType.NotEqual, fieldName, valuesAsObj));
			return this;
		}
		public EsqExt<T> AddInFilter(string fieldName, params object[] values)
		{
			Filters.Add(CreateFilterWithParameters(FilterComparisonType.Equal, fieldName, values));
			return this;
		}
		public EsqExt<T> AddNotInFilter(string fieldName, params object[] values)
		{
			Filters.Add(CreateFilterWithParameters(FilterComparisonType.NotEqual, fieldName, values));
			return this;
		}
		public EsqExt<T> AddFilterWithParameters(string fieldName, object value, FilterComparisonType type = FilterComparisonType.Equal)
		{
			Filters.Add(CreateFilterWithParameters(type, fieldName, value));
			return this;
		}
		public EsqExt<T> AddFilter(IEntitySchemaQueryFilterItem filterItem)
		{
			Filters.Add(filterItem);
			return this;
		}
		public EsqExt<T> AddFilter(string fieldName, string otherFieldName, FilterComparisonType type = FilterComparisonType.Equal)
		{
			Filters.Add(CreateFilter(type, fieldName, otherFieldName));
			return this;
		}
		public EsqExt<T> AddIsNullFilter(string fieldName)
		{
			Filters.Add(CreateIsNullFilter(fieldName));
			return this;
		}
		public EsqExt<T> AddIsNotNullFilter(string fieldName)
		{
			Filters.Add(CreateIsNotNullFilter(fieldName));
			return this;
		}
		public EntitySchemaQueryFilterCollection CreateNewFilterGroup(LogicalOperationStrict logicalOperation)
		{
			var filterGroup = new EntitySchemaQueryFilterCollection(this, logicalOperation);
			return filterGroup;
		}
		public EntitySchemaQueryFilterCollection CreateNewFilterGroup(LogicalOperationStrict logicalOperation, params IEntitySchemaQueryFilterItem[] filters)
		{
			var filterGroup = CreateNewFilterGroup(logicalOperation);
			if (filters != null || filters.Length > 0)
				filterGroup.AddRange(filters);
			return filterGroup;
		}
		public EsqExt<T> AddFilterGroup(LogicalOperationStrict logicalOperation, params IEntitySchemaQueryFilterItem[] filters)
		{
			var filterGroup = CreateNewFilterGroup(logicalOperation, filters);
			AddFilterGroup(filterGroup);
			return this;
		}
		public EsqExt<T> AddFilterGroup(EntitySchemaQueryFilterCollection filterGroup)
		{
			Filters.Add(filterGroup);
			return this;
		}
		#endregion
	}
}

File: EsqExt.js
class EsqExt {
	constructor(data){
		if (typeof data === "string" || data instanceof String)
			this.esq = new Terrasoft.EntitySchemaQuery(data);
		else if (data instanceof Terrasoft.EntitySchemaQuery)
			this.esq = data;
		else
			throw new Error(`Esq constructor not implement data: ${data} with type: ${typeof data}`);
	}
	get filters() {
		return this.esq.filters;
	}
	get clientCache() {
		return this.esq.clientESQCacheParameters.cacheItemName;
	}
	set clientCache(value) {
		this.esq.clientESQCacheParameters.cacheItemName = value;
	}
	get serverCache() {
		return this.esq.serverESQCacheParameters;
	}
	set serverCache(value) {
		this.esq.serverESQCacheParameters = value;
	}
	addColumn(columnPath, alias){
		if (!alias) alias = columnPath;
		return this.esq.addColumn(columnPath, alias);
	}
	addColumns(...columnPaths){
		for(let columnPath of columnPaths)
		this.addColumn(columnPath);
	}
	getEntity(primaryColumnValue){
			return new Promise(resolve =>
			this.esq.getEntity(primaryColumnValue, result => resolve(result), this)
		);
	}
	getEntityCollection() {
		return new Promise(resolve =>
			this.esq.getEntityCollection(result => resolve(result), this)
		);
	}
	getFirstEntity() {
		this.esq.rowCount = 1;
		return new Promise(resolve =>
			this.esq.getEntityCollection(result => resolve(result.collection.first()), this)
		);
	}
	addColumnFilter(columnPath1, columnPath2, comparisonType = Terrasoft.ComparisonType.EQUAL){
		let filter = Terrasoft.createFilter(comparisonType, columnPath1, columnPath2);
		this.filters.addItem(filter);
	}
	addFilterWithParameter(columnPath, value, comparisonType = Terrasoft.ComparisonType.EQUAL) {
		let filter = Terrasoft.createColumnFilterWithParameter(comparisonType, columnPath, value);
		this.filters.addItem(filter);
	}
	addInFilterWithParameters(columnPath, inValues) {
		let filter = Terrasoft.createColumnInFilterWithParameters(columnPath, inValues);
		this.filters.addItem(filter);
	}
	addNotInFilterWithParameters(columnPath, notInValues) {
		let filter = Terrasoft.createColumnInFilterWithParameters(columnPath, notInValues);
		filter.comparisonType = Terrasoft.ComparisonType.NOT_EQUAL;
		this.filters.addItem(filter);
	}
	addIsNullFilter(columnPath) {
		let filter = Terrasoft.createColumnIsNullFilter(columnPath)
		this.filters.addItem(filter);
	}
	addIsNotNullFilter(columnPath) {
		let filter = Terrasoft.createColumnIsNotNullFilter(columnPath)
		this.filters.addItem(filter);
	}
}
